File: flutter/lib/data/message_repository.dart
import '../models/message.dart';
import 'mock_messages.dart';

/// Read and write access to conversations.
abstract interface class MessageRepository {
  /// Every message for [contactId], oldest first.
  Future<List<Message>> fetchThread(String contactId);

  /// Appends an outgoing message from the user and returns the stored copy.
  Future<Message> send({required String contactId, required String body});

  /// Appends a CareConnect alert to the thread and returns the stored copy.
  ///
  /// Declared separately from [send] rather than as an optional argument on it,
  /// because the two have genuinely different meanings: one is the user
  /// speaking, the other is the app recording something it did.
  Future<Message> recordAlert({
    required String contactId,
    required String body,
  });
}

/// In-memory implementation seeded from [buildMockMessages].
///
/// Messages sent during a session are kept for the lifetime of the process,
/// which is enough for the Week 4 demo: leaving a thread and coming back shows
/// what was sent.
class MockMessageRepository implements MessageRepository {
  MockMessageRepository({DateTime? now, List<Message>? seed})
      : _clock = now ?? DateTime.now(),
        _messages = List<Message>.of(
          seed ?? buildMockMessages(now ?? DateTime.now()),
        );

  final DateTime _clock;
  final List<Message> _messages;
  int _sequence = 0;

  @override
  Future<List<Message>> fetchThread(String contactId) async {
    final List<Message> thread = _messages
        .where((Message message) => message.contactId == contactId)
        .toList()
      ..sort((Message a, Message b) => a.sentAt.compareTo(b.sentAt));
    return thread;
  }

  @override
  Future<Message> send({
    required String contactId,
    required String body,
  }) async {
    return _append(
      contactId: contactId,
      body: body,
      author: MessageAuthor.me,
      kind: MessageKind.text,
    );
  }

  @override
  Future<Message> recordAlert({
    required String contactId,
    required String body,
  }) async {
    return _append(
      contactId: contactId,
      body: body,
      author: MessageAuthor.system,
      kind: MessageKind.alert,
    );
  }

  Message _append({
    required String contactId,
    required String body,
    required MessageAuthor author,
    required MessageKind kind,
  }) {
    _sequence++;
    final Message message = Message(
      id: 'sent-$_sequence',
      contactId: contactId,
      author: author,
      kind: kind,
      body: body.trim(),
      // The seeded clock keeps ordering stable in tests; a real backend would
      // stamp this server-side.
      sentAt: _clock.add(Duration(minutes: _sequence)),
      status: DeliveryStatus.sent,
    );
    _messages.add(message);
    return message;
  }
}

File: mobile/src/data/messageRepository.ts
import { chronological, type Message } from '../models/message';
import { buildMockMessages } from './mockMessages';

/** Read and write access to conversations. */
export interface MessageRepository {
  /** Every message for a contact, oldest first. */
  fetchThread(contactId: string): Promise<Message[]>;

  /** Appends an outgoing message from the user and returns the stored copy. */
  send(contactId: string, body: string): Promise<Message>;

  /**
   * Appends a CareConnect alert and returns the stored copy.
   *
   * Declared separately from `send` rather than as an optional argument on it,
   * because the two have genuinely different meanings: one is the user
   * speaking, the other is the app recording something it did.
   */
  recordAlert(contactId: string, body: string): Promise<Message>;
}

export interface MockMessageRepositoryOptions {
  now?: number;
  seed?: Message[];
  /** How many `fetchThread` calls fail before the repository starts working. */
  failures?: number;
}

/**
 * In-memory implementation seeded from the prototype's conversations.
 *
 * Messages sent during a session are kept for the lifetime of the process,
 * which is enough for the Week 5 demo: leaving a conversation and coming back
 * shows what was sent.
 */
export function createMockMessageRepository(
  options: MockMessageRepositoryOptions = {},
): MessageRepository {
  const clock = options.now ?? Date.now();
  const messages: Message[] = [...(options.seed ?? buildMockMessages(clock))];
  const failures = options.failures ?? 0;
  let attempts = 0;
  let sequence = 0;

  const append = (
    contactId: string,
    body: string,
    author: Message['author'],
    kind: Message['kind'],
  ): Message => {
    sequence += 1;
    const message: Message = {
      id: `sent-${sequence}`,
      contactId,
      author,
      kind,
      body: body.trim(),
      // The seeded clock keeps ordering stable in tests; a real backend would
      // stamp this server-side.
      sentAt: clock + sequence * 60_000,
      status: 'sent',
    };
    messages.push(message);
    return message;
  };

  return {
    async fetchThread(contactId) {
      attempts += 1;
      if (attempts <= failures) throw new Error('offline');
      return chronological(messages.filter((message) => message.contactId === contactId));
    },
    async send(contactId, body) {
      return append(contactId, body, 'me', 'text');
    },
    async recordAlert(contactId, body) {
      return append(contactId, body, 'system', 'alert');
    },
  };
}
